File: ext/webidl/benches/dict.js
// Copyright 2018-2025 the Deno authors. MIT license.

// deno-lint-ignore-file

import {
  converters,
  createDictionaryConverter,
} from "ext:deno_webidl/00_webidl.js";

const TextDecodeOptions = createDictionaryConverter(
  "TextDecodeOptions",
  [
    {
      key: "stream",
      converter: converters.boolean,
      defaultValue: false,
    },
  ],
);
globalThis.TextDecodeOptions = TextDecodeOptions;

// Sanity check
{
  const o = TextDecodeOptions(undefined);
  if (o.stream !== false) {
    throw new Error("Unexpected stream value");
  }
}

function handwrittenConverter(V) {
  const defaultValue = { stream: false };
  if (V === undefined || V === null) {
    return defaultValue;
  }
  if (V.stream !== undefined) {
    defaultValue.stream = !!V.stream;
  }
  return defaultValue;
}
globalThis.handwrittenConverter = handwrittenConverter;

File: ext/webidl/benches/dict.rs
// Copyright 2018-2025 the Deno authors. MIT license.

use deno_bench_util::bench_js_sync;
use deno_bench_util::bench_or_profile;
use deno_bench_util::bencher::Bencher;
use deno_bench_util::bencher::benchmark_group;
use deno_core::Extension;

fn setup() -> Vec<Extension> {
  deno_core::extension!(
    deno_webidl_bench,
    esm_entry_point = "ext:deno_webidl_bench/setup.js",
    esm = ["ext:deno_webidl_bench/setup.js" = "benches/dict.js"]
  );

  vec![deno_webidl::deno_webidl::init(), deno_webidl_bench::init()]
}

fn converter_undefined(b: &mut Bencher) {
  bench_js_sync(b, r#"TextDecodeOptions(undefined);"#, setup);
}

fn handwritten_baseline_undefined(b: &mut Bencher) {
  bench_js_sync(b, r#"handwrittenConverter(undefined)"#, setup);
}

fn converter_object(b: &mut Bencher) {
  bench_js_sync(b, r#"TextDecodeOptions({});"#, setup);
}

fn handwritten_baseline_object(b: &mut Bencher) {
  bench_js_sync(b, r#"handwrittenConverter({})"#, setup);
}

benchmark_group!(
  benches,
  converter_undefined,
  handwritten_baseline_undefined,
  converter_object,
  handwritten_baseline_object,
);
bench_or_profile!(benches);
